relay-server/src/actors/processor.rs

```diff
@@ -2795,14 +2795,20 @@ impl Service for EnvelopeProcessorService {
                 return;
             };
 
+            let mut shutdown = false;
+
             loop {
                 let next_msg = async { tokio::join!(rx.recv(), semaphore.clone().acquire_owned()) };
 
                 tokio::select! {
                    biased;
 
-                    // TODO(iker): deal with the error when the sender of the channel is dropped.
-                    _ = subscription.changed() => self.global_config = subscription.borrow().clone(),
+                    update = subscription.changed(), if !shutdown => {
+                        match update {
+                            Ok(()) => self.global_config = subscription.borrow().clone(),
+                            Err(_) => shutdown = true,
+                        };
```

Comment:
Are the changes to this file actually worth it? What's the damage without this change? One additional clone of the global config at the end of the lifecycle?

Reply:
When the global config service gets the shutdown signal it immediately shuts down and drops all the memory it owns, including the sender of the channel. When the sender is dropped, reading a new subscription always errors, and that results in the processor not handling new messages and thus dropping data (not enough time to process in-flight envelopes). Shutting the processor down on the shutdown signal before consuming the remaining messages has the same effect.

This is reproducible with `tests/integration/test_basic.py::test_graceful_shutdown` -- waiting for 2 seconds after sending the shutdown signal is enough to process the data, but not waiting for it results in data loss.

I thought about adding a comment about this, but I didn't want to write too much and assumed the clarification on the `Subscription` message in the global config service was good enough. Let me know if you feel otherwise and I'll add it.

Reply:
Added in https://github.com/getsentry/relay/pull/2445/commits/6a1d346a079573594a99f8d16e2b3989e09d1c19.

Reply:
Simplified in [d2cf59a](https://github.com/getsentry/relay/pull/2445/commits/d2cf59a763797fa0d63a38bd68afe8c57c6f3aea).